applications/asset_tracker/src/main.c

```diff
@@ -1354,22 +1573,38 @@ void main(void)
 		}
 
 		if ((fds[0].revents & POLLNVAL) == POLLNVAL) {
-			if (atomic_get(&reconnect_to_cloud)) {
-				k_delayed_work_cancel(&cloud_reboot_work);
-				LOG_INF("Attempting reconnect...");
+			if (atomic_get(&cloud_association) ==
+					CLOUD_ASSOCIATION_STATE_RECONNECT) {
+				reconnecting = true;
+				reconnect_delay_s = 10;
 				goto connect;
 			}
 			LOG_ERR("Socket error: POLLNVAL");
 			LOG_ERR("The cloud socket was unexpectedly closed.");
 			error_handler(ERROR_CLOUD, -EIO);
-			return;
 		}
 
 		if ((fds[0].revents & POLLHUP) == POLLHUP) {
+			if ((atomic_get(&cloud_connect_attempts) == 1) &&
+				(atomic_get(&cloud_association) ==
+				 CLOUD_ASSOCIATION_STATE_INIT)) {
+				LOG_INF("Disconnected by the cloud.");
+				LOG_INF("This can occur during provisioning or when the device has invalid certificates.");
+#if defined(CONFIG_LWM2M_CARRIER)
+				LOG_INF("Rebooting...");
+				/* Reconnect does not work with LWM2M */
+				k_sleep(K_SECONDS(10));
+				LOG_PANIC();
+				sys_reboot(0);
```

Comment:
I think we should use `cloud_reboot_handler(...)` or `error_handler(...)` here instead to keep cloud's reboot handling easier to maintain